tapfeatures/aux_channel_negotiator.go

```diff
@@ -0,0 +1,165 @@
+	// required feature is not supported by the remote peer we have to
+	// return an error and drop the connection. Whether we support all of
+	// the remote required features is a responsibility of the remote peer.
+	// If we fail to support a remotely required feature they are the ones
+	// to drop the connection (by returning an error right here).
+	err = checkRequiredBits(getLocalFeatureVec(), peerVec)
+	if err != nil {
+		return err
+	}
+
+	// Store this peer's features.
+	n.peerFeatures.Store(peer, peerVec)
+
+	return nil
+}
+
+// GetReestablishFeatures is called when sending a channel_reestablish message.
+// It returns feature bits based on the specific channel identified by its
+// funding outpoint and aux channel blob.
+func (n *AuxChannelNegotiator) GetReestablishFeatures(
+	_ wire.OutPoint, _ tlv.Blob) (tlv.Blob, error) {
+
+	var buf bytes.Buffer
+
+	// We could store per-channel features, but currently there's no need to
+	// support that. We report the same feature vector over all of our
+	// custom channels.
+	features := getLocalFeatureVec()
+	err := features.Encode(&buf)
+	if err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+// ProcessReestablishFeatures handles received channel_reestablish feature TLVs.
+// This is a blocking call - the channel link will wait for this method to
+// complete before continuing channel operations. The implementation can modify
+// aux channel behavior based on the negotiated features.
+func (n *AuxChannelNegotiator) ProcessReestablishFeatures(
+	fundingPoint wire.OutPoint, features tlv.Blob,
+	auxChanBlob tlv.Blob) error {
+
+	buf := bytes.NewBuffer(features)
+	peerVec := lnwire.NewRawFeatureVector()
+	err := peerVec.Decode(buf)
+	if err != nil {
+		return err
+	}
+
+	// Before we store this channel's supported features we need to first
+	// check if our required features are supported by the remote party. If
+	// a locally required feature is not supported we have to return an
+	// error and drop the connection. Whether we support all of the remote
+	// required features is a responsibility of the remote peer. If we fail
+	// to support a remotely required feature they are the ones to drop the
+	// connection (by returning an error right here).
+	err = checkRequiredBits(getLocalFeatureVec(), peerVec)
+	if err != nil {
+		return err
+	}
+
+	n.chanFeatures.Store(fundingPoint, peerVec)
+
+	return nil
+}
+
+// GetPeerFeatures returns the negotiated custom channel type that was
+// established with the given peer.
+func (n *AuxChannelNegotiator) GetPeerFeatures(
+	peer route.Vertex) *lnwire.FeatureVector {
+
+	rawfeatures, ok := n.peerFeatures.Load(peer)
+	if !ok {
+		rawfeatures = lnwire.NewRawFeatureVector()
+	}
+
+	return lnwire.NewFeatureVector(rawfeatures, featureNames)
+}
+
+// checkRequiredBits is a helper method that checks if all of the required bits
+// of the first vector are supported by the second vector.
+func checkRequiredBits(a, b *lnwire.RawFeatureVector) error {
```

Comment:
This one would be more immediately clear if 'a' and 'b' were replaced by 'local' and 'peer' or something, per the error message (a cute idiom I've seen before is 'ours' and 'hers' 😄).